### PR TITLE
Update example with zio-http and zio-json

### PR DESCRIPTION
**Pull request by mvelimir**

### CODE REVIEW

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala [hunk 1]
```diff
@@ -1,46 +1,37 @@
...
-    implicit def encoder[A: Encoder]: EntityEncoder[AppTask, A] = jsonEncoderOf[AppTask, A]
-
-    HttpRoutes.of[AppTask] { case GET -> Root / "statuses" =>
+  def statuses(backendUri: Uri, service: ZLayer[Clock, Throwable, Clock with OpenTracing]): HttpApp[Clock, Throwable] = {
+    HttpApp.collectM { case Method.GET -> Root / "statuses" =>
       val zio =
         for {
           _       <- OpenTracing.tag(Tags.SPAN_KIND.getKey, Tags.SPAN_KIND_CLIENT)
-          _       <- OpenTracing.tag(Tags.HTTP_METHOD.getKey, GET.name)
+          _       <- OpenTracing.tag(Tags.HTTP_METHOD.getKey, "GET")
```
Comment (drmarjanovic) [inline opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala]: Is there any constant for this?

**Review by drmarjanovic — COMMENTED** [review]
LGTM.

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/Status.scala [hunk 2]
```diff
@@ -1,15 +1,13 @@
...
-import io.circe.generic.semiauto._
+import zio.json._
 
 final case class Status(name: String, status: String)
 
 object Status {
-  implicit val decoder: Decoder[Status] = deriveDecoder
-  implicit val encoder: Encoder[Status] = deriveEncoder
+  implicit val decoder: JsonDecoder[Status] = DeriveJsonDecoder.gen[Status]
+  implicit val encoder: JsonEncoder[Status] = DeriveJsonEncoder.gen[Status]
```
Comment (mijicd) [inline opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/Status.scala]: You can use `DeriveJsonCodec` instead.

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusService.scala [hunk 3]
```diff
@@ -1,34 +1,22 @@
...
-  val dsl: Http4sDsl[AppTask] = Http4sDsl[AppTask]
-  import dsl._
-
-  implicit def encoder[A: Encoder]: EntityEncoder[AppTask, A] = jsonEncoderOf[AppTask, A]
-
-  def status(service: ZLayer[Clock, Throwable, Clock with OpenTracing]): HttpRoutes[AppTask] =
-    HttpRoutes.of[AppTask] { case request @ GET -> Root / "status" =>
-      val headers = request.headers.toList.map(h => h.name.value -> h.value).toMap
+  def status(service: ZLayer[Clock, Throwable, Clock with OpenTracing]): HttpApp[Clock, Throwable] =
+    Http.collectM { case request@Method.GET -> Root / "status" =>
```
Comment (mijicd) [inline opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusService.scala]: Super nitpick: I'd make a line break here as it's more readable imho. Something like:
```suggestion
    Http.collectM { 
      case request@Method.GET -> Root / "status" =>
        // stuff you alread have
    }
```

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/Statuses.scala [hunk 4]
```diff
@@ -1,11 +1,10 @@
...
-import io.circe.generic.semiauto._
+import zio.json._
 
 final case class Statuses(data: List[Status]) extends AnyVal
 
 object Statuses {
-  implicit val decoder: Decoder[Statuses] = deriveDecoder
-  implicit val encoder: Encoder[Statuses] = deriveEncoder
+  implicit val decoder: JsonDecoder[Statuses] = DeriveJsonDecoder.gen[Statuses]
+  implicit val encoder: JsonEncoder[Statuses] = DeriveJsonEncoder.gen[Statuses]
```
Comment (mijicd) [inline opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/Statuses.scala]: Use `DeriveJsonCodec`.

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala [hunk 5]
```diff
@@ -1,46 +1,37 @@
...
-
-  def statuses(backendUri: Uri, service: ZLayer[Clock, Throwable, Clock with OpenTracing]): HttpRoutes[AppTask] = {
-    val dsl: Http4sDsl[AppTask] = Http4sDsl[AppTask]
-    import dsl._
-
-    implicit def encoder[A: Encoder]: EntityEncoder[AppTask, A] = jsonEncoderOf[AppTask, A]
-
-    HttpRoutes.of[AppTask] { case GET -> Root / "statuses" =>
+  def statuses(backendUri: Uri, service: ZLayer[Clock, Throwable, Clock with OpenTracing]): HttpApp[Clock, Throwable] = {
+    HttpApp.collectM { case Method.GET -> Root / "statuses" =>
```
Comment (mijicd) [inline opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala]: Same as above.

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala (hunk 1 above)
Comment (mijicd) [reply opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala]: There's no need for this change.

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusService.scala [hunk 6]
```diff
@@ -1,34 +1,22 @@
...
+      val headers = request.headers.map(h => h.name.toString -> h.value.toString).toMap
       ZIO.unit
         .spanFrom(HttpHeadersFormat, new TextMapAdapter(headers.asJava), "/status")
-        .provideLayer(service) *> Ok(ServiceStatus.up("backend").asJson)
+        .provideLayer(service) *> ZIO.succeed(Response.jsonString(ServiceStatus.up("backend").toJson))
```
Comment (mijicd) [inline opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusService.scala]:
```suggestion
        .provideLayer(service)
        .as(Response.jsonString(ServiceStatus.up("backend").toJson))
```

Alternatively, push the `provideLayer` to the end as it "reads better":

```suggestion
        .as(Response.jsonString(ServiceStatus.up("backend").toJson))
        .provideLayer(service).
```

File: project/Dependencies.scala [hunk 7]
```diff
@@ -41,22 +43,24 @@ object Dependencies {
...
-    "dev.zio"                      %% "zio-config"                    % Versions.zioConfig,
-    "dev.zio"                      %% "zio-config-magnolia"           % Versions.zioConfig,
-    "dev.zio"                      %% "zio-config-typesafe"           % Versions.zioConfig,
-    "dev.zio"                      %% "zio-interop-cats"              % Versions.zioInteropCats
+    "org.typelevel"                 %% "cats-core"                     % "2.6.1",
+    "io.circe"                      %% "circe-generic"                 % "0.14.1",
+    "org.http4s"                    %% "http4s-core"                   % Versions.http4s,
+    "org.http4s"                    %% "http4s-blaze-server"           % Versions.http4s,
+    "org.http4s"                    %% "http4s-dsl"                    % Versions.http4s,
+    "org.http4s"                    %% "http4s-circe"                  % Versions.http4s,
```
Comment (mijicd) [inline project/Dependencies.scala:51]: These should not be needed anymore.

File: project/Dependencies.scala [hunk 8]
```diff
@@ -41,22 +43,24 @@ object Dependencies {
...
+    "io.jaegertracing"               % "jaeger-core"                   % Versions.jaeger,
+    "io.jaegertracing"               % "jaeger-client"                 % Versions.jaeger,
+    "io.jaegertracing"               % "jaeger-zipkin"                 % Versions.jaeger,
+    "com.github.pureconfig"         %% "pureconfig"                    % "0.16.0",
```
Comment (mijicd) [inline project/Dependencies.scala:55]: Same here.

**Review by mijicd — CHANGES_REQUESTED** [review]

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala (hunk 1 above)
Comment (mvelimir) [reply opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala]: `GET` was defined in http4s, and the `GET` object in zio-http doesn't seem to have a `name` member. I could either use the `toString` method, or `asJHttpMethod.name`, both of which return the name of the HTTP method.

File: project/Dependencies.scala (hunk 7 above)
Comment (mvelimir) [reply project/Dependencies.scala:51]: The opentelemetry example is still using them. Should I move them to `opentelemetryExample` for the time being?

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusService.scala (hunk 3 above)
Comment (mvelimir) [reply opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusService.scala]: It does look cleaner.

File: project/Dependencies.scala (hunk 8 above)
Comment (mvelimir) [reply project/Dependencies.scala:55]: `JaegerTracer` still depends on the jaegertracing library. Am I supposed to use something else instead?

File: project/Dependencies.scala (hunk 7 above)
Comment (mijicd) [reply project/Dependencies.scala:51]: I see, leave it for now then.

File: project/Dependencies.scala (hunk 8 above)
Comment (mijicd) [reply project/Dependencies.scala:55]: I'm referring to pureconfig, not to the tracer. If it's used in other example, you can remove it later.

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala (hunk 1 above)
Comment (mijicd) [reply opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala]: You may use the one from sttp then.

File: opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala [hunk 9]
```diff
@@ -16,28 +17,29 @@ import scala.jdk.CollectionConverters._
...
+            _       <- OpenTracing.inject(HttpHeadersFormat, buffer)
+            headers <- extractHeaders(buffer)
+            up       = Status.up("proxy")
+            res     <- Client
+                         .status(backendUri.withPath("status"), headers)
+                         .map(_.body)
+                         .flatMap {
+                           case Right(s) => ZIO.succeed(Response.jsonString(Statuses(List(s, up)).toJson))
+                           case _        => ZIO.succeed(Response.jsonString(Statuses(List(Status.down("backend"), up)).toJson))
+                         }
```
Comment (mijicd) [inline opentracing-example/src/main/scala/zio/telemetry/opentracing/example/http/StatusesService.scala]: `jsonString` is pure, therefore you can simplify:

```suggestion
                         .map { res =>
                           val status = res.body.getOrElse(Status.down("backend"))
                           val statuses =  Statuses(List(status, up))
                           Response.jsonString(statuses.toJson)
                         }
```

**Review by mijicd — CHANGES_REQUESTED** [review]

**Review by mijicd — APPROVED** [review]

**Pull request merged**